ocaml/xapi/test_vm.ml

```diff
@@ -0,0 +1,100 @@
+(*
+ * Copyright (C) 2017 Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+open Stdext
+open OUnit
+open Test_highlevel
+
+module VMSetBiosStrings = Generic.Make (Generic.EncapsulateState (struct
+  module Io = struct
+    type input_t = (API.vm_bios_string_keys * string) list
+    type output_t = (exn, (string * string) list) Either.t
+    let string_of_input_t = fun x -> Printf.sprintf "%s"
+      (String.concat "; " (List.map (fun (k,v) -> (Record_util.vm_bios_key_to_string k) ^ "=" ^ v) x))
+    let string_of_output_t = Test_printers.(either exn (assoc_list string string))
+  end
+  module State = Test_state.XapiDb
+
+  let name_label = "a"
+
+  let load_input __context _ =
+    ignore (Test_common.make_vm ~__context ~name_label ())
+
+  let extract_output __context value =
+    let self = List.hd (Db.VM.get_by_name_label ~__context ~label:name_label) in
+    try
+      Xapi_vm.set_bios_strings ~__context ~self ~value;
+      Either.Right (Db.VM.get_bios_strings ~__context ~self)
+    with e -> Either.Left e
+
+  let big_str = String.make (Xapi_globs.bios_string_limit_size + 1) 'x'
+  let non_printable_str1 = Printf.sprintf "xyz%c" (Char.chr 31)
+  let non_printable_str2 = Printf.sprintf "xyz%c" (Char.chr 127)
+  let bios_str1 = [`biosvendor, "Test"; `biosversion, "Test Inc. A08"]
+  let bios_str2 = [`systemmanufacturer, "Test Inc."; `systemproductname, "Test bios strings"; `systemversion, "8.1.1 SP1 build 8901"]
+  let bios_str3 = [`enclosureassettag, "testassettag12345"]
+
+  let get_bios_strings input =
```

Comment:
Please don't use a function like this to "compute" the output of the function-under-test. You really have to just hardcode the expected output. I know it may seem counterintuitive, but this is what unit tests are supposed to do.

Reply:
Updated, Thanks